### PR TITLE
AuthSystem Combined for DEMO

### LINKED ISSUE #66: < Auth System > Set Up Authentication API Routes: AuthRoutes
- **Description**: Define the API routes for handling authentication actions, including login, registration, and logout. The routes should handle requests from the frontend to authenticate users, validate credentials, and issue tokens for session management. Additionally, include routes for handling the user’s session, such as logging out and token revocation.
- **Tag**: `#66`
- **Owner**: Haiyi
- **Expected Outcome**: A set of routes that facilitate login, registration, reset passsword, and logout operations for user authentication.

### LINKED ISSUE #67: < Auth System > Set Up Authentication Controller
- **Description**: Implement the business logic for user authentication, including handling login, registration, and logout operations. This controller will interact with the _UserModel_ to validate credentials, hash passwords for secure storage, and manage tokens for user sessions. It will also handle errors related to invalid credentials, missing fields, or failed login attempts.
- **Tag**: `#67`
- **Owner**: Haiyi
- **Expected Outcome**: A controller that manages authentication logic, including verifying user credentials, generating tokens, and managing user sessions.

### LINKED ISSUE #68: < Auth System > Implement User Authentication Interface: AuthPage
- **Description**: Implement the frontend page for user authentication, including both login and registration forms. This page should handle user inputs, validate the required fields, and send requests to the backend for login and registration actions. The page should also handle error messages ( invalid credentials, missing fields) and provide appropriate feedback to the user.
- **Tag**: `#68`
- **Owner**: Haiyi
- **Expected Outcome**: Functional registration and login forms that communicate with the back-end.

### LINKED ISSUE #72: < Auth System > Set Up UserModel
**Description**: Create a `UserModel' in Sequelize with attributes for `userId`, `email`, `hashedPassword`, and `role`. Add model-level validation for required fields and email format.
- **Tag**: `#72`
- **Owner**: Haiyi
- **Expected Outcome**: A functioning Sequelize model stored in SQLite, with appropriate validations.

### LINKED ISSUE #73: < Auth System > Middleware for Route Protection: AuthMiddleware
- **Description**: Implement middleware to protect routes that require authentication. This middleware checks for a valid session or token before allowing access to protected endpoints. If the user is not authenticated, the middleware will return an [ 401 Unauthorized] error and prevent further access to the route.
- **Tag**: `#73`
- **Owner**: Haiyi
- **Expected Outcome**: Middleware that verifies authentication and blocks unauthorized access.

### LINKED ISSUE #74: < Auth System > Front-End Integration for Registration and Login
- **Description**: Add front-end forms for user registration and login. Integrate these forms with the corresponding back-end routes using fetch or Axios for API calls.
- **Tag**: `#74`
- **Owner**: Haiyi
- **Expected Outcome**: Functional registration and login forms that communicate with the back-end.

### LINKED ISSUE #75: < Auth System > Middleware for Route Protection: RoleMiddleware
- **Description**: Implement middleware to protect routes that require specific user roles. This middleware checks the authenticated user's role and compares it against the required role for the route. If the user does not have the correct role, access will be blocked, and a [ 403 Forbidden ] error will be returned.
- **Tag**: `#75`
- **Owner**: Haiyi
- **Expected Outcome**: Middleware that checks the user's role and blocks access to routes for unauthorized roles.

### PR DESCRIPTION
**Pull request by HaiyiCai**
Here's the latest version for DEMO, will debug after review.

### CODE REVIEW

File: backend/controllers/AuthController.js [hunk 1]
```diff
@@ -1 +1,318 @@
...
+import UserModel  from "../models/UserModel.js";
+import nodemailer from "nodemailer";
+import { v4 as uuidv4 } from "uuid";
+
+// Environment Variables
+const JWT_SECRET = process.env.JWT_SECRET || "your_secret_key";
+
+// Includes logic for managing user authentication, including:
+/* 1. Registering a new user:  Handles user registration with required fields validations & password security. */
+export async function registerUser(req, res) {  // TODO Change name as register()
```
Comment (wnauda) [inline backend/controllers/AuthController.js:27]: This method looks good, works as intended in my testing.

File: backend/controllers/AuthController.js [hunk 2]
```diff
@@ -1 +1,318 @@
...
+    res.status(200).json({ message: " ><> Email verified successfully. <><" });
+  } catch (error) {
+    console.error("Error during email verification:", error);
+    res.status(500).json({ error: "Internal server error."} );
+  }
+} 
+
+
+/* 3. Log in a user: Verifies credentials, generates authentication tokens, and responds with user details.*/
+export async function login(req, res) {
```
Comment (wnauda) [inline backend/controllers/AuthController.js:117]: Works as intended in my testing, looks good

File: backend/controllers/AuthController.js [hunk 3]
```diff
@@ -1 +1,318 @@
...
+    });
+  } catch (error) {
+    // Handles invalid login attempts with appropriate error messages.
+    console.error("Error during login:", error);
+    res.status(500).json({ error: "Internal server error." });
+  }
+}
+
+/* 4. Log out a user: Invalidates tokens to ensure users can securely log out. */
+export async function logout(req, res) {
```
Comment (wnauda) [inline backend/controllers/AuthController.js:170]: working!

File: backend/controllers/AuthController.js [hunk 4]
```diff
@@ -1 +1,318 @@
...
+    const mailOptions = {
+      from: process.env.EMAIL_USER,
+      to: user.email,
+      subject: "><> Password Reset Request",
+      text: `Click the link to reset your password: ${resetLink}. This link will expire in 1 hour.`,
+    };
+    console.log("Mail options created.");
+
+    console.log("Sending email through transporter...");
+    await transporter.sendMail(mailOptions);
```
Comment (wnauda) [inline backend/controllers/AuthController.js:241]: this call is erroring out:
Error during password reset request: Error: Invalid login: 535-5.7.8 Username and Password not accepted. For more information, go to
535 5.7.8  https://support.google.com/mail/?p=BadCredentials 6a1803df08f44-6d8da9fe7f5sm8021006d6.80 - gsmtp

File: backend/middleware/AuthMiddleware.js [hunk 5]
```diff
@@ -1 +1,64 @@
...
+// **Expected Outcome**:  Ensures user has valid session/ token and blocks unauthenticated access.
+
+/* Works with UserModel.js to retrieve and validate user data as needed. It's used in routes that 
+require user authentication: profile updates, placing orders, accessing user-specific data, etc. 
+*/
+import jwt from "jsonwebtoken";
+import UserModel from "../models/UserModel.js";
+
+// Validates user credentials by checking the authentication token or session associated with the incoming request.
+const authMiddleware = async (req, res, next) => {
```
Comment (wnauda) [inline backend/middleware/AuthMiddleware.js:16]: this is working now!

File: backend/middleware/RoleMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  return (req, res, next) => {
+    try{
+      if(!req.user){                  // Ensure AuthMiddleware has attached a user object to the request.
+        return res.status(401).json({error: 'No user found, access denied.'});
+      }
+/* 1. Role Verification:
+  - Checks the roles attribute in the user object attached by AuthMiddleware.js.
+  - Confirms the user has the required role(s) for the route.
+*/
+      const userRole = req.user.roles;      // Extract role from authenticated user.
```
Comment (wnauda) [inline backend/middleware/RoleMiddleware.js:25]: "req.user" is the user's id not a User object. You need to call getUserById() before you can retrieve the roles

File: frontend/main/app/AppController.js [hunk 7]
```diff
@@ -26,29 +31,40 @@ export class AppController {
        sellProductsPage: new SellProductsPage(),
      };
 
-     // Set the default page as the marketplace
-     this.#currentView = this.#views.marketplace;
+     // Set the initial view to AuthPage if not logged in, else default page is the marketplace
+     this.#currentView = authService.isLoggedIn() ? this.#views.marketplace : this.#views.auth;
    }
 
-   async render(id=null) {
```
Comment (wnauda) [inline frontend/main/app/AppController.js:33]: The id parameter is still required to navigate to specific product pages, that change should be reversed.

**Review by wnauda — CHANGES_REQUESTED** [review]
A few changes are needed, but overall you're on the right track!

File: frontend/main/app/AppController.js [hunk 8]
```diff
@@ -6,17 +6,22 @@ import { ProductPage } from '../components/ProductPage/ProductPage.js';
 import { NavigationMenu } from '../components/NavigationMenu/NavigationMenu.js';
 import { ProfilePage } from '../components/ProfilePage/ProfilePage.js';
 import { SellProductsPage } from '../components/MarketplacePage/SellProductsPage.js';
+import { AuthPage } from '../components/AuthPage/AuthPage.js';
+import { authService } from '../services/AuthService.js'; // Handles user authentication state
```
Comment (wnauda) [inline frontend/main/app/AppController.js:10]: this is not a valid import. AuthService.js does not provide an export named authService, only the class AuthService. you should create an instance of the AuthService class in AuthService.js and export it if you want to use it like this.

File: frontend/main/components/AuthPage/AuthPage.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    The AuthPage provides the user interface for interacting with the authentication system. 
+    It will only function properly once the backend routes and logic are implemented.
+  *Logic:
+    - Create UI forms for login, registration, and password reset.
+    - Use API calls to connect with the AuthRoutes.
+    - Dependencies: The AuthPage interacts with the AuthRoutes to send and receive data for authentication processes.
+*/
+
+import { BaseComponent } from '../../app/BaseComponent.js';
+import { authService } from '../../services/AuthService.js';
```
Comment (wnauda) [inline frontend/main/components/AuthPage/AuthPage.js:50]: same problem with authService as in AppController.js

File: frontend/main/components/AuthPage/AuthPage.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+          headers: { 'Content-Type': 'application/json' },
+          body: JSON.stringify({ email, password }),
+        });
+
+        if (!response.ok) {
+          throw new Error('Login failed');
+        }
+
+        const { token, userId } = await response.json();
+        authService.login(token, userId);
```
Comment (wnauda) [inline frontend/main/components/AuthPage/AuthPage.js:145]: When I tested the login, it was successful on the backend but this call is failing and showing a failure alert.

File: frontend/main/app/AppController.js [hunk 11]
```diff
@@ -62,15 +78,22 @@ export class AppController {
...
-     if(viewName === 'productPage'){
-      this.render(params.prodid);
-     }else{
-      this.render();
+    if (!this.#views[viewName]) {
+      throw new Error(`View "${viewName}" not found.`);
+    }
+
+     // Restrict access to certain views based on authentication
+    if(!authService.isLoggedIn() && (viewName !== 'auth' || 'marketplace' )){
```
Comment (wnauda) [inline frontend/main/app/AppController.js]: || 'marketplace' doesn't do anything here, 'marketplace' will always evaluate to true.

File: frontend/main/services/AuthService.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            this.user = null;
+        }
+    }
+
+    isLoggedIn() {
+        //  Check if user is defined and not null, while token is not expired
+        return !!this.user && !this.isTokenExpired(this.user.token); 
+    }
+
+    isTokenExpired() {
```
Comment (wnauda) [inline frontend/main/services/AuthService.js]: this needs to have a parameter called token

File: backend/controllers/AuthController.js [hunk 13]
```diff
@@ -1 +1,318 @@
...
+    const token = jwt.sign(
+      tokenPayload, 
+      JWT_SECRET, 
+      { expiresIn: "1h" }
+    );
+    console.log("JWT generated successfully:", token);
+
+    res.status(200).json({
+      message: " ><> Login successful. <><",
+      token,
```
Comment (wnauda) [inline backend/controllers/AuthController.js:159]: add a userId field with the value validCredentials.userId here

**Comment by HaiyiCai** [comment]
Thnak you so much for helping me debugging and fixing the issue, Winifred!! I really appreciate your effort and contribution to Auth Sytem, we can't do this without you.

**Comment by batterydied** [comment]
Hello, Winifred.
Check discord for my messages

**Pull request closed without merging**